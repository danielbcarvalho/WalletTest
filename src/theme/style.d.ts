import 'styled-components/native';

declare module 'styled-components/native' {
  interface TipographyKindProps {
    fontFamily: string;
    fontSize: string;
    lineHeight?: string;
  }

  interface TipographyProps {
    h1: TipographyKindProps;
    h2: TipographyKindProps;
    h3: TipographyKindProps;
    h4: TipographyKindProps;
    h5: TipographyKindProps;
    p: TipographyKindProps;
    pSmall: TipographyKindProps;
  }

  interface LabelProps extends TipographyKindProps {
    color: string;
  }
  interface InputProps extends TipographyKindProps {
    color: string;
    backgroundColor: string;
    borderRadius: string;
    padding: string;
    label: LabelProps;
  }

  interface ButtonProps extends TipographyKindProps {
    radius: string;
    primary: {
      backgroundColor: string;
      color: string;
    };
    secondary: {
      backgroundColor: string;
      color: string;
    };
    disabled: {
      backgroundColor: string;
      color: string;
    };
  }

  interface PalleteProps {
    primary: {
      main: string;
    };
    success: {
      main: string;
    };
    alert: {
      main: string;
    };
    warning: {
      main: string;
    };
    info: {
      main: string;
    };
    background: {
      default: string;
      overlay: string;
    };
    textOnBackground: {
      default: string;
      overlay: string;
    };
  }

  interface Header {
    title: HeaderTitle;
  }

  interface HeaderTitle extends TipographyKindProps {
    register: {
      color: string;
    };
    cardList: {
      color: string;
    };
  }

  export interface DefaultTheme {
    pallete: PalleteProps;
    typography: TipographyProps;
    components: {
      button: ButtonProps;
      input: InputProps;
      header: Header;
    };
  }
}
