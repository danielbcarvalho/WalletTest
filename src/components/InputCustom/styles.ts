import styled from 'styled-components/native';
import MaskInput from 'react-native-mask-input';

interface ContainerProps {
  width?: number;
}

export const Container = styled.View<ContainerProps>`
  width: ${({ width }) => (width ? `${width}%` : '100%')};
  margin: 16px 10px;
`;

export const Label = styled.Text`
  font-size: ${({ theme }) => theme.components.input.label.fontSize};
  font-family: ${({ theme }) => theme.components.input.label.fontFamily};
  color: ${({ theme }) => theme.components.input.label.color};
  margin-bottom: 6px;
`;

export const MaskInputStyled = styled(MaskInput)`
  padding: 0 16px;
  margin-top: 4px;
  height: 45px;
  background-color: ${({ theme }) => theme.components.input.backgroundColor};
  color: ${({ theme }) => theme.components.input.color};
  border-radius: 6px;
  overflow: hidden;
  width: 100%;
  font-size: ${({ theme }) => theme.components.input.fontSize};
  font-family: ${({ theme }) => theme.components.input.fontFamily};
`;
