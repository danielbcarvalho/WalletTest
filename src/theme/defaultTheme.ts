import { DefaultTheme } from 'styled-components/native';

export const defaultTheme: DefaultTheme = {
  pallete: {
    primary: {
      main: '#142995',
    },
    success: {
      main: '#A5FF32',
    },
    alert: {
      main: '#FF0000',
    },
    warning: {
      main: '#FFB800',
    },
    info: {
      main: '#12C2E9',
    },
    background: {
      default: '#142995',
      overlay: '#EEEEEE',
    },
    textOnBackground: {
      default: '#FFF',
      overlay: '#142995',
    },
  },
  typography: {
    h1: {
      fontFamily: 'PTSans-Caption',
      fontSize: '28px',
      lineHeight: '42px',
    },
    h2: {
      fontFamily: 'PTSans-Caption',
      fontSize: '26px',
      lineHeight: '38px',
    },
    h3: {
      fontFamily: 'PTSansCaption',
      fontSize: '22px',
      lineHeight: '34px',
    },
    h4: {
      fontFamily: 'PTSans-Regular',
      fontSize: '20px',
      lineHeight: '32px',
    },
    h5: {
      fontFamily: 'PTSans-Regular',
      fontSize: '18px',
      lineHeight: '30px',
    },
    p: {
      fontFamily: 'PTSans-Regular',
      fontSize: '16px',
      lineHeight: '28px',
    },
    pSmall: {
      fontFamily: 'PTSans-Regular',
      fontSize: '14px',
      lineHeight: '26px',
    },
  },
  components: {
    button: {
      radius: '12px',
      fontFamily: 'PTSans-Regular',
      fontSize: '18px',
      lineHeight: '28px',
      primary: {
        backgroundColor: '#12C2E9',
        color: '#FFFFFF',
      },
      secondary: {
        backgroundColor: '#A5FF32',
        color: '#142995',
      },
      disabled: {
        backgroundColor: '#EEEEEE',
        color: '#BBBBBB',
      },
    },
    input: {
      label: {
        fontFamily: 'PTSans-Regular',
        fontSize: '14px',
        color: '#FFFFFF',
        lineHeight: '16px',
      },
      fontFamily: 'PTSans-Regular',
      fontSize: '16px',
      color: '#111',
      backgroundColor: '#EEEEEE',
      borderRadius: '12px',
      padding: '16px',
    },
    header: {
      title: {
        fontFamily: 'PTSans-Caption',
        fontSize: '22px',
        lineHeight: '24px',
        register: {
          color: '#142995',
          backgroundColor: '#FFFFFF',
        },
        cardList: {
          color: '#12C2E9',
        },
      },
    },
    card: {
      black: {
        backgroundColor: '#000000',
        color: '#FFFFFF',
      },
      green: {
        backgroundColor: '#A5FF32',
        color: '#3F3F3F',
      },
      title: {
        fontFamily: 'PTSans-Regular',
        fontSize: '18px',
        lineHeight: '20px',
      },
      name: {
        fontFamily: 'PTSans-Regular',
        fontSize: '16px',
        lineHeight: '18px',
      },
      text: {
        fontFamily: 'PTSans-Regular',
        fontSize: '14px',
        lineHeight: '16px',
      },
    },
  },
};

export const mockDefaultTheme: object = defaultTheme;
