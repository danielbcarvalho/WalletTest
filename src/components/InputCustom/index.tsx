import React, { useCallback, useState } from 'react';
import { TextInputProps } from 'react-native';

import { Container, Label, MaskInputStyled } from './styles';
import { cardFormValidationSchema } from '../Card/CardRegistrationForm/schema';

interface Props extends TextInputProps {
  mask?: any;
  name: string;
  label: string;
  value: string;
  width?: number;
}

function InputCustom({ mask, value, label, width, name, ...props }: Props) {
  const [isValid, setValid] = useState(true);

  const validateField = useCallback(() => {
    const fieldValidationSchema = (cardFormValidationSchema.fields as any)[
      name
    ];

    fieldValidationSchema
      .validate(value)
      .then(() => {
        setValid(true);
      })
      .catch(() => {
        setValid(false);
      });
  }, [name, value]);

  return (
    <Container width={width}>
      <Label>{label}</Label>
      <MaskInputStyled
        mask={mask}
        value={value}
        isDirty={!isValid}
        autoComplete="off"
        autoCorrect={false}
        accessibilityLabel={label}
        onSelectionChange={validateField}
        {...props}
      />
    </Container>
  );
}

export default InputCustom;
