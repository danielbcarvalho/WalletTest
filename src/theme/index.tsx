import React, { createContext, useMemo } from 'react';
import { ThemeProvider as ThemeProviderStyled } from 'styled-components/native';

import { defaultTheme } from './defaultTheme';

export enum Theme {
  light = 'light',
  dark = 'dark',
}

interface IApplication {
  children: JSX.Element;
}

const themes = {
  light: defaultTheme,
};

export const ThemeContext = createContext({
  theme: Theme.light,
});

export function ThemeProvider({ children }: IApplication) {
  const theme = Theme.light;

  const memoizedValue = useMemo(() => ({ theme }), [theme]);

  return (
    <ThemeContext.Provider value={memoizedValue}>
      <ThemeProviderStyled theme={themes[theme]}>
        {children}
      </ThemeProviderStyled>
    </ThemeContext.Provider>
  );
}
