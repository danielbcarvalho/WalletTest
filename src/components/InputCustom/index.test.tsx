import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

import InputCustom from './';
import { ThemeProvider } from '../../theme';

describe('InputCustom', () => {
  test('renders correctly', () => {
    const label = 'card number';
    const value = '1';
    const onChangeText = jest.fn();

    const { getByLabelText, getByText, toJSON } = render(
      <ThemeProvider>
        <InputCustom
          label={label}
          value={value}
          onChangeText={onChangeText}
          name="number"
        />
      </ThemeProvider>,
    );

    const inputElement = getByLabelText(label);
    expect(inputElement).toHaveProp('value', value);

    expect(toJSON()).toMatchSnapshot();
  });

  test('validates field on selection change', () => {
    const label = 'Card Number';
    const value = '1234567890';
    const validateFieldMock = jest.fn();

    const { getByLabelText } = render(
      <ThemeProvider>
        <InputCustom
          label={label}
          value={value}
          onSelectionChange={validateFieldMock}
        />
      </ThemeProvider>,
    );

    const inputElement = getByLabelText(label);
    fireEvent(inputElement, 'selectionChange');

    expect(validateFieldMock).toHaveBeenCalledTimes(1);
  });
});
