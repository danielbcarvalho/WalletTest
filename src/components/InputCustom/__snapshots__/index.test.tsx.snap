// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputCustom renders correctly 1`] = `
<View
  style={
    {
      "marginBottom": 16,
      "marginLeft": 10,
      "marginRight": 10,
      "marginTop": 16,
      "width": "100%",
    }
  }
>
  <Text
    style={
      {
        "color": "#FFFFFF",
        "fontFamily": "PTSans-Regular",
        "fontSize": 14,
        "marginBottom": 6,
      }
    }
  >
    card number
  </Text>
  <TextInput
    accessibilityLabel="card number"
    autoComplete="off"
    autoCorrect={false}
    isDirty={false}
    onChangeText={[Function]}
    onSelectionChange={[Function]}
    style={
      {
        "backgroundColor": "#EEEEEE",
        "borderBottomColor": "transparent",
        "borderBottomLeftRadius": 6,
        "borderBottomRightRadius": 6,
        "borderColor": "transparent",
        "borderLeftColor": "transparent",
        "borderRightColor": "transparent",
        "borderStyle": "solid",
        "borderTopColor": "transparent",
        "borderTopLeftRadius": 6,
        "borderTopRightRadius": 6,
        "borderWidth": 2,
        "color": "#111",
        "fontFamily": "PTSans-Regular",
        "fontSize": 16,
        "height": 45,
        "marginTop": 4,
        "overflow": "hidden",
        "paddingBottom": 0,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 0,
        "width": "100%",
      }
    }
    value="1"
  />
</View>
`;
